### PR TITLE
Add basic rate limiting to backup

### PR DESCRIPTION
**Pull request by ghost**
Introduces the --rate-limit flag for the backup subcommand. This sets a
package global indicating the KB/s rate for uploads.

Tested with iftop

References #1216

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -72,3 +72,7 @@
 [[constraint]]
   branch = "master"
   name = "golang.org/x/sys"
+
+[[constraint]]
```
Comment (fd0) [inline Gopkg.toml]: Hm, did you add this yourself? The way I understood `dep` works is that it should figure out a dependency by inspecting the code, so this entry shouldn't be needed.

**Review by fd0 — CHANGES_REQUESTED** [review]
This looks really good already! However, I'd like to have a discussion on how to integrate it.

**Comment by fd0** [comment]
At the moment this PR conflicts with #1276, which also changes the way the `http.Transport` is passed to the backends, so we need to resolve that once #1276 (which is basically ready) is merged.

Then I'd like to talk about:
 * Integrating this with the `sftp` and `local` backends (yes, also for the `local` backend, in case people use that with other fuse-based file systems like `sshfs` or `cifs`)
 * Getting rid of the global variables in the `backend` package
 * I'd like to do the same for download
 * Finding a better name for the option, `--upload-limit`? And add a separate option for download

It feels to me that the HTTP Transport, while used for several backends, is the wrong place for the limiter. Maybe it should be a backend config thing, so that each backend takes care of it (and we can still provide a wrapper around `http.Transport`).

What do you think?

File: Gopkg.toml (hunk 1 above)
Comment (rmdashrf) [reply Gopkg.toml]: Yeah, this is unnecessary. Still getting used to `dep` -- I forget that the Gopkg.lock file is checked in and will contain this revision anyway.

**Comment by rmdashrf** [comment]
@fd0 - I agree with all of your points. I whipped this up quickly because my B2 backup was killing my internet speed, so it was mainly a proof of concept. I think Golang even explicity states that `http.RoundTripper` implementations should not mess with the request. :)

Without globals, I think the best way to approach this is to define something like
```
type Limiter interface {
   LimitReader(r io.Reader) io.Reader
   LimitWriter(r io.Writer) io.Writer
}
```
which would be initialized by the CLI and passed as parameters to all of the backends. The backend implementations would be responsible for wrapping the readers/writers when appropriate.

I have no strong opinions on the command flag, although I do prefer using "limit" as the prefix (e.g. `--limit-upload` and `--limit-download`) for easier copy-pasta. :)

**Comment by fd0** [comment]
Good idea with the interface, but I think defining that for an `io.Reader` is sufficient. We'll see that during the implementation. I'd rather like to differentiate between up and downstream, e.g. like this:
```go
type Limiter interface {
    Upstream(io.Reader) io.Reader
    Downstream(io.Reader) io.Reader
}
```

**Comment by fawick** [comment]
@rmdashrf (Sorry for side-tracking this issue discussion) Out of curiosity, did you consider using golang.org/x/time/rate, too? If so, what are the drawbacks when compared to the juju package?

**Comment by rmdashrf** [comment]
@fawick The juju implementation already implements a tested rate limited `io.Reader`. It seems like the `golang.org/x/time/rate` package would require more work and more testing.

**Comment by fawick** [comment]
I see, thank you!

**Comment by fd0** [comment]
Hm, I've had a look, and it works well for upstream but fails for downstream with the sftp backend. Any ideas?

File: internal/limiter/limiter.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package limiter
+
+import (
+	"io"
+)
+
+type Limiter interface {
```
[the remark this answers is not in the document]
Comment (fd0) [reply internal/limiter/limiter.go]: Please add a short description here.

File: internal/limiter/static_limiter.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		upstreamBucket   *ratelimit.Bucket
+		downstreamBucket *ratelimit.Bucket
+	)
+
+	if uploadKb > 0 {
+		upstreamBucket = ratelimit.NewBucketWithRate(toByteRate(uploadKb), int64(toByteRate(uploadKb)))
+	}
+
+	if downloadKb > 0 {
+		upstreamBucket = ratelimit.NewBucketWithRate(toByteRate(downloadKb), int64(toByteRate(downloadKb)))
```
Comment (fd0) [inline internal/limiter/static_limiter.go]: Here's a bug: That should be `downstreamBucket` instead of `upstreamBucket`, then it works.

**Review by fd0 — CHANGES_REQUESTED** [review]

File: internal/limiter/static_limiter.go (hunk 3 above)
Comment (rmdashrf) [reply internal/limiter/static_limiter.go]: /facepalm

**Review by fd0 — APPROVED** [review]

**Comment by fd0** [comment]
Thanks!

**Comment by zcalusic** [comment]
This is really great work, @rmdashrf! 

Even if I don't have immediate need for it, I know the time will come when it'll be indispensable, 'cause it already happened multiple times in the past with rsync, wget, and other apps... IMHO, it's a great feature for a backup program.

Thank you for your contribution!

**Pull request merged**